registry/internal/tools/finder/github.go

```diff
@@ -107,21 +124,68 @@ type githubCommit struct {
 	} `json:"commit"`
 }
 
-func (gc githubCommit) Pseudo() (string, error) {
-	ts, err := time.Parse(time.RFC3339, gc.Commit.Author.Date)
+func (gc githubCommit) Pseudo(tags []Tag, hasGomod bool) (string, error) {
+	naked, semver, err := gc.nakedPseudo(tags)
 	if err != nil {
 		return "", err
 	}
 
-	date := ts.Format("200601020300")
+	if semver == nil || semver.Major < 2 || hasGomod {
+		return naked, nil
+	}
+	return naked + "+incompatible", nil
+}
 
-	pseudo := fmt.Sprintf(
-		"v0.0.0-%s-%s+incompatible",
-		date,
-		gc.SHA[0:12], // what Go does
-	)
+func (gc githubCommit) nakedPseudo(tags []Tag) (string, *SemVer, error) {
+	ts, err := time.Parse(time.RFC3339, gc.Commit.Author.Date)
+	if err != nil {
+		return "", nil, err
+	}
+
+	date := ts.Format("20060102150405")
+	shortSHA := gc.SHA[0:12] // what Go does
+
+	if len(tags) == 0 {
+		return fmt.Sprintf("v0.0.0-%s-%s", date, shortSHA), nil, nil
+	}
+
+	lastVersion := tags[0].SemVer
+	semver := parseSemVer(lastVersion)
+
+	if semver.isPre {
+		// TODO should this always be ".0" or do we increment the pre version?
+		return fmt.Sprintf("%s.0.%s-%s", lastVersion, date, shortSHA), semver, nil
+	}
+
+	return fmt.Sprintf("v%d.%d.%d-0.%s-%s", semver.Major, semver.Minor, semver.Patch+1, date, shortSHA), semver, nil
+}
 
-	return pseudo, nil
+var semVerRe = regexp.MustCompile(`^v(\d+)(?:\.(\d+)(?:\.((\d+)(-pre)?))?)?$`)
+
+// -rc is commonly used, but not in the spec
+//var semVerRe = regexp.MustCompile(`^v(\d+)(?:\.(\d+)(?:\.(\d+(-pre|-rc)?))?)?$`)
+
+func parseSemVer(semver string) *SemVer {
+	matches := semVerRe.FindStringSubmatch(semver)
+	if matches == nil {
+		return nil
+	}
+	if matches[2] == "" && matches[3] == "" {
+		return &SemVer{unsafeIToA(matches[1]), 0, 0, false}
+	}
+	if matches[3] == "" {
+		return &SemVer{unsafeIToA(matches[1]), unsafeIToA(matches[2]), 0, false}
+	}
+	return &SemVer{unsafeIToA(matches[1]), unsafeIToA(matches[2]), unsafeIToA(matches[4]), matches[5] != ""}
+}
+
+// only call this if you are sure that s is convertible to an int
+func unsafeIToA(s string) int {
```

Comment:
using the library, we'll check the error (bool) on Parse, and we should handle it rather than panic

Reply:
even with an intentionally unexported method, where the code shouldn't be reached?

we know that these are integers because we filter tags based on a regex. I don't feel that strongly about this either, but it will make the code more annoying to read.

Reply:
I removed this bit of code entirely.